cachelib/allocator/CacheAllocator-inl.h

```diff
@@ -2763,12 +2633,118 @@ void CacheAllocator<CacheTrait>::evictForSlabRelease(
   }
 }
 
+template <typename CacheTrait>
+typename CacheAllocator<CacheTrait>::WriteHandle
+CacheAllocator<CacheTrait>::advanceIteratorAndTryEvictRegularItem(
+    MMContainer& mmContainer, EvictionIterator& itr) {
+  // we should flush this to nvmcache if it is not already present in nvmcache
+  // and the item is not expired.
+  Item& item = *itr;
+  const bool evictToNvmCache = shouldWriteToNvmCache(item);
+
+  auto token = evictToNvmCache ? nvmCache_->createPutToken(item.getKey())
+                               : typename NvmCacheT::PutToken{};
+
+  // If there are other accessors, we should abort. Acquire a handle here since
+  // if we remove the item from both access containers and mm containers
+  // below, we will need a handle to ensure proper cleanup in case we end up
+  // not evicting this item
+  auto evictHandle = accessContainer_->removeIf(item, &itemExclusivePredicate);
+  if (!evictHandle) {
+    ++itr;
+    stats_.evictFailAC.inc();
+    return evictHandle;
+  }
+
+  mmContainer.remove(itr);
+  XDCHECK_EQ(reinterpret_cast<uintptr_t>(evictHandle.get()),
+             reinterpret_cast<uintptr_t>(&item));
+  XDCHECK(!evictHandle->isInMMContainer());
+  XDCHECK(!evictHandle->isAccessible());
+
+  // Invalidate iterator since later on if we are not evicting this
+  // item, we may need to rely on the handle we created above to ensure
+  // proper cleanup if the item's raw refcount has dropped to 0.
+  // And since this item may be a parent item that has some child items
+  // in this very same mmContainer, we need to make sure we drop this
+  // exclusive iterator so we can gain access to it when we're cleaning
+  // up the child items
+  itr.destroy();
+
+  // Ensure that there are no accessors after removing from the access
+  // container
+  XDCHECK(evictHandle->getRefCount() == 1);
+
+  if (evictToNvmCache && shouldWriteToNvmCacheExclusive(item)) {
+    XDCHECK(token.isValid());
```

Comment:
This failed our internal tests. You can repro this by running this cachebench test config `cachelib/cachebench/test_configs/consistency/navy.json`. The issue is related to comments at L2644.

```
F1013 14:57:11.677257   633 CacheAllocator-inl.h:2677] Check failed: token.isValid() 
*** Aborted at 1665698231 (Unix time, try 'date -d @1665698231') ***
*** Signal 6 (SIGABRT) (0x6300000161) received by PID 353 (pthread TID 0x7f2645c00640) (linux TID 633) (maybe from PID 353, UID 99) (code: -6), stack trace: ***
    @ 000000000698b4be folly::symbolizer::(anonymous namespace)::signalHandler(int, siginfo_t*, void*)
                       ./folly/experimental/symbolizer/SignalHandler.cpp:449
    @ 0000000000000000 (unknown)
    @ 000000000009c9d3 __GI___pthread_kill
    @ 00000000000444ec __GI_raise
    @ 000000000002c432 __GI_abort
    @ 00000000069ac304 folly::LogCategory::admitMessage(folly::LogMessage const&) const
                       ./folly/logging/LogCategory.cpp:71
    @ 00000000069aa42d folly::LogStreamProcessor::logNow()
                       ./folly/logging/LogStreamProcessor.cpp:190
    @ 00000000069aa91c folly::LogStreamVoidify<true>::operator&(std::ostream&)
                       ./folly/logging/LogStreamProcessor.cpp:222
    @ 0000000005c608b6 facebook::cachelib::CacheAllocator<facebook::cachelib::LruCacheTrait>::advanceIteratorAndTryEvictRegularItem(facebook::cachelib::MMLru::Container<facebook::cachelib::CacheItem<facebook::cachelib::LruCacheTrait>, &facebook::cachelib::CacheItem<facebook::cachelib::LruCacheTrait>::mmHook_>&, facebook::cachelib::MMLru::Container<facebook::cachelib::CacheItem<facebook::cachelib::LruCacheTrait>, &facebook::cachelib::CacheItem<facebook::cachelib::LruCacheTrait>::mmHook_>::Iterator&)
                       ./cachelib/allocator/CacheAllocator-inl.h:2677
                       -> ./cachelib/allocator/CacheAllocator.cpp
    @ 0000000005c5810d facebook::cachelib::CacheAllocator<facebook::cachelib::LruCacheTrait>::findEviction(signed char, signed char)
                       ./cachelib/allocator/CacheAllocator-inl.h:1251
                       -> ./cachelib/allocator/CacheAllocator.cpp
    @ 0000000005c5897c facebook::cachelib::CacheAllocator<facebook::cachelib::LruCacheTrait>::allocateChainedItemInternal(facebook::cachelib::detail::ReadHandleImpl<facebook::cachelib::CacheItem<facebook::cachelib::LruCacheTrait> > const&, unsigned int)
                       ./cachelib/allocator/CacheAllocator-inl.h:409
                       -> ./cachelib/allocator/CacheAllocator.cpp
    @ 0000000005c8bee0 facebook::cachelib::CacheAllocator<facebook::cachelib::LruCacheTrait>::allocateChainedItem(facebook::cachelib::detail::ReadHandleImpl<facebook::cachelib::CacheItem<facebook::cachelib::LruCacheTrait> > const&, unsigned int)
                       ./cachelib/allocator/CacheAllocator-inl.h:381
                       -> ./cachelib/allocator/CacheAllocator.cpp
    @ 00000000059c0e0a facebook::cachelib::cachebench::CacheStressor<facebook::cachelib::CacheAllocator<facebook::cachelib::LruCacheTrait> >::stressByDiscreteDistribution(facebook::cachelib::cachebench::ThroughputStats&)
                       ./cachelib/cachebench/cache/Cache-inl.h:355
                       -> ./cachelib/cachebench/runner/Stressor.cpp
    @ 00000000000df4e4 execute_native_thread_routine
    @ 000000000009ac0e start_thread
    @ 000000000012d1db __clone3

```

Reply:
Fixed.